From a4bccc1eb19ed0ec2c1f789dbe58d1476bcc8449 Mon Sep 17 00:00:00 2001
From: CIQ Kernel Automation <ciq_kernel_automation@ciq.com>
Date: Mon, 27 Apr 2026 07:11:38 +0000
Subject: [PATCH] mISDN: hfcpci: Fix use-after-free bug in hfcpci_softirq

jira VULN-61254
cve-bf CVE-2021-47356
commit-author Duoming Zhou <duoming@zju.edu.cn>
commit 175302f6b79ebbb207c2d58d6d3e679465de23b0

The function hfcpci_softirq() is a timer handler. If it
is running, the timer_pending() will return 0 and the
del_timer_sync() in HFC_cleanup() will not be executed.
As a result, the use-after-free bug will happen. The
process is shown below:

    (cleanup routine)          |        (timer handler)
HFC_cleanup()                  | hfcpci_softirq()
 if (timer_pending(&hfc_tl))   |
   del_timer_sync()            |
 ...                           | ...
 pci_unregister_driver(hc)     |
  driver_unregister            |  driver_for_each_device
   bus_remove_driver           |   _hfcpci_softirq
    driver_detach              |   ...
     put_device(dev) //[1]FREE |
                               |    dev_get_drvdata(dev) //[2]USE

The device is deallocated is position [1] and used in
position [2].

Fix by removing the "timer_pending" check in HFC_cleanup(),
which makes sure that the hfcpci_softirq() have finished
before the resource is deallocated.

Fixes: 009fc857c5f6 ("mISDN: fix possible use-after-free in HFC_cleanup()")
	Signed-off-by: Duoming Zhou <duoming@zju.edu.cn>
	Signed-off-by: David S. Miller <davem@davemloft.net>
(cherry picked from commit 175302f6b79ebbb207c2d58d6d3e679465de23b0)
	Signed-off-by: CIQ Kernel Automation <ciq_kernel_automation@ciq.com>
---
 drivers/isdn/hardware/mISDN/hfcpci.c | 3 +--
 1 file changed, 1 insertion(+), 2 deletions(-)

diff --git a/drivers/isdn/hardware/mISDN/hfcpci.c b/drivers/isdn/hardware/mISDN/hfcpci.c
index 6f66a389e741..0a5fdbdd71d4 100644
--- a/drivers/isdn/hardware/mISDN/hfcpci.c
+++ b/drivers/isdn/hardware/mISDN/hfcpci.c
@@ -2345,8 +2345,7 @@ HFC_init(void)
 static void __exit
 HFC_cleanup(void)
 {
-	if (timer_pending(&hfc_tl))
-		del_timer_sync(&hfc_tl);
+	del_timer_sync(&hfc_tl);
 
 	pci_unregister_driver(&hfc_driver);
 }
-- 
2.47.3

